### PR TITLE
Main page data from backend

### PR DESCRIPTION
**Pull request by duckmoll**
Rebase to main branch.
Video was posted in slack.

### CODE REVIEW

File: NewDawn/NewDawn/MainPageViewController.swift [hunk 1]
```diff
@@ -38,4 +40,75 @@ class MainPageViewController: UIViewController {
...
+    
+    func readReponse(response: NSDictionary) -> Void{
+        let profile_responses = response["objects"] as? [[String: Any]]
+        for profile in profile_responses!{
+            let dummy_user = getProfileInfo(profile: profile)
+            USER_DUMMY_DATA.append(dummy_user as NSDictionary)
+        }
+    }
+
+    func encodeParams(raw_params: Dictionary<String, String>) -> String{
```
Comment (duckmoll) [inline NewDawn/NewDawn/MainPageViewController.swift]: However, I see HttpUtil extends from UIImageView though. Can I still do that?

File: NewDawn/NewDawn/MainPageViewController.swift [hunk 2]
```diff
@@ -38,4 +40,75 @@ class MainPageViewController: UIViewController {
...
+                    "media": "media/images/testcat.JPG",
+                    "caption": "First image"
+                ],
+            ],
+            "question_answers":[String]()
+            ] as [String : Any]
+        for answer_question in (profile["answer_question"] as? [[String: Any]])!{
+            let answer_question_dict = [
+                "id": answer_question["order"]!,
+                QUESTION: answer_question[QUESTION]!,
```
Comment (duckmoll) [inline NewDawn/NewDawn/MainPageViewController.swift]: I will change QUESTION and ANSWER.
QUESTION_ANSWERS: front end defines it "question_answers", but back end defines it "answer_question"
ID: ID is a little too big for picture ids as global name

File: NewDawn/NewDawn/MainPageViewController.swift [hunk 3]
```diff
@@ -16,17 +16,19 @@ class MainPageViewController: UIViewController {
 
     override func viewDidLoad() {
         super.viewDidLoad()
-        if ProfileIndexUtil.loadProfileIndex() >= USER_DUMMY_DATA.count {
-            // TODO: When the user has seen all profiles, we go back to the first profile.
-            // In the future, we should go to an ending page
-            self.localStoreKeyValue(key: MAIN_PAGE_PROFILE_INDEX, value: 0)
+        self.getUserData(){ () -> () in
```
Comment (duckmoll) [inline NewDawn/NewDawn/MainPageViewController.swift]: It seems that each time user click skip, it will re-fetch profiles from backend. Ideally we should only fetch once when the main page is on daily fresh.
You can do that by moving this function to checkMainPageReload() in ViewerController.swift, and store the entire list of UserProfiles into local storage. You will then only need to read profiles from local storage.

Comment (duckmoll) [reply NewDawn/NewDawn/MainPageViewController.swift]: Will ViewController be too heavy if I move all fetching user data logic to it?

Comment (tangziyi001) [reply NewDawn/NewDawn/MainPageViewController.swift]: Yes. But we can refactor that later

Comment (tangziyi001) [reply NewDawn/NewDawn/MainPageViewController.swift]: Ideally, the processSessionTask should be in HttpUtil also. If you want, You can copy that function to HttpUtil for now. 

Then You can try put getUserData, getProfileRequest etc. into UserProfile class as a static method. Create another static method called buildUserProfile, which takes a request as input, and then return a list of UserProfile object as output. Then you use that function like UserProfile.buildUserProfile(request..) in changeMainPageReload() to fetch user info and make them as UserProfile objects and store them into local storage.

**Comment by tangziyi001** [comment]
Generally LGTM. If you want to make the code cleaner, you can follow the comment above.

File: NewDawn/NewDawn/MainPageViewController.swift (hunk 3 above)
Comment (duckmoll) [reply NewDawn/NewDawn/MainPageViewController.swift]: Where do you want to put the UserProfileBuilder class?

Comment (tangziyi001) [reply NewDawn/NewDawn/MainPageViewController.swift]: UserModel.swift would be fine

File: NewDawn/NewDawn/Models/UserModels.swift [hunk 4]
```diff
@@ -77,7 +77,7 @@ let USER_DUMMY_DATA_2: NSDictionary = [
     ]
 ]
 
-let USER_DUMMY_DATA = [USER_DUMMY_DATA_1, USER_DUMMY_DATA_2]
+var USER_DUMMY_DATA = [USER_DUMMY_DATA_1, USER_DUMMY_DATA_2]
```
Comment (tangziyi001) [inline NewDawn/NewDawn/Models/UserModels.swift]: revert this change

File: NewDawn/NewDawn/Models/UserModels.swift [hunk 5]
```diff
@@ -167,3 +167,78 @@ class UserProfile {
...
+            let dummy_user = parseProfileInfo(profile_data: profile)
+            let dummy_user_profile = UserProfile(data: dummy_user as NSDictionary)
+            fetched_users.append(dummy_user_profile)
+        }
+        let profilesData = try! JSONEncoder().encode(fetched_users)
+        UserDefaults.standard.set(profilesData, forKey: "UserProfiles")
+    }
+    
+    
+    static func getUserProfileListFromLocalStorage()->[UserProfile]{
```
Comment (tangziyi001) [inline NewDawn/NewDawn/Models/UserModels.swift]: You might want to check if the list of user profile exists. If not then it needs to call fetchAndStoreInLocalStorage() first.

**Comment by duckmoll** [comment]
It does not process the request for some reason..

**Comment by tangziyi001** [comment]
Try print something and see where goes wrong?

**Comment by duckmoll** [comment]
HttpUtil.processSessionTasks seems not working, and it simply skips that process without any error.

File: NewDawn/NewDawn/Models/UserModels.swift [hunk 6]
```diff
@@ -167,3 +167,85 @@ class UserProfile {
...
+        let profilesData = UserDefaults.standard.data(forKey: "UserProfiles")
+        var profilesArray = [UserProfile]()
+        if profilesData == nil{
+            UserProfileBuilder.fetchAndStoreUserProfiles()
```
Comment (tangziyi001) [inline NewDawn/NewDawn/Models/UserModels.swift]: As mentioned above, even after this function returns, the processSessionTask might not be finished yet since it is running async. So you cannot be able to get local storage data ready immediately.

I'll try to find how to fix it tonight

Comment (duckmoll) [reply NewDawn/NewDawn/Models/UserModels.swift]: Sure thing.

**Comment by tangziyi001** [comment]
So it seems that a simple way to make the "wait for processSessionTask to complete" is to use a semaphore. You need to make some functions to be non-static and add a semaphore attribute to the class. Then use it to notify the getUserProfileListFromLocalStorage() function if the fetchAndStore is completed. Specifically:

1. Make fetchAndStoreUserProfiles, fetchAndStoreInLocalStorage, getUserProfileListFromLocalStorage to be non-static functions.
2. Add a semaphore into the UserProfileBuilder class:
```
 var semaphore: DispatchSemaphore
 init() {
        semaphore = DispatchSemaphore(value: 0)
 }
```
3. Add semaphore.signal() to the end of fetchAndStoreInLocalStorage()
4. Add semaphore.wait() in getUserProfileListFromLocalStorage() right after calling fetchAndStoreUserProfiles()

After making these changes, the code for reading local storage in getUserProfileListFromLocalStorage() function will hang forever until fetchAndStoreInLocalStorage() is completed. That's to say you have to make sure that this fetchAndStoreInLocalStorage callback must be called and completed successfully. 

To avoid hanging the program forever, you can add a timeout to the semaphore.

**Comment by tangziyi001** [comment]
Or you can keep the static functions and use ` static var semaphore = DispatchSemaphore(value: 0)` instead.

File: NewDawn/NewDawn/MainPageViewController.swift [hunk 7]
```diff
@@ -16,12 +16,16 @@ class MainPageViewController: UIViewController {
 
     override func viewDidLoad() {
         super.viewDidLoad()
-        if ProfileIndexUtil.loadProfileIndex() >= USER_DUMMY_DATA.count {
+        if TimerUtil.isOutdated(){
+            UserProfileBuilder.fetchAndStoreUserProfiles()
```
Comment (tangziyi001) [inline NewDawn/NewDawn/MainPageViewController.swift]: Does this solve the problem?

Comment (duckmoll) [reply NewDawn/NewDawn/MainPageViewController.swift]: I realize that my getUserProfileListFromLocalStorage does not need that process session logic.
But that semaphore technique is really awesome. Everything works fine as of this commit.

File: NewDawn/NewDawn/Models/UserModels.swift [hunk 8]
```diff
@@ -167,3 +167,85 @@ class UserProfile {
...
+    static func fetchAndStoreInLocalStorage(response: NSDictionary)-> Void{
+        var fetched_users = [UserProfile]()
+        let profile_responses = response["objects"] as? [[String: Any]]
+        for profile in profile_responses!{
+            let dummy_user = parseProfileInfo(profile_data: profile)
+            let dummy_user_profile = UserProfile(data: dummy_user as NSDictionary)
+            fetched_users.append(dummy_user_profile)
+        }
+        let profilesData = try! JSONEncoder().encode(fetched_users)
+        UserDefaults.standard.set(profilesData, forKey: "UserProfiles")
```
[the remark this answers is not in the document]
Comment (tangziyi001) [reply NewDawn/NewDawn/Models/UserModels.swift]: Try use LocalStorageUtil

File: NewDawn/NewDawn/Models/UserModels.swift [hunk 9]
```diff
@@ -167,3 +167,82 @@ class UserProfile {
...
+    }
+    
+    
+    static func getUserProfileListFromLocalStorage()->[UserProfile]{
+        let profilesData = UserDefaults.standard.data(forKey: "UserProfiles")
```
Comment (tangziyi001) [inline NewDawn/NewDawn/Models/UserModels.swift]: Same here

File: NewDawn/NewDawn/MainPageViewController.swift (hunk 7 above)
Comment (duckmoll) [reply NewDawn/NewDawn/MainPageViewController.swift]: > Better to use LocalStorageUtil function to do this instead of encoding and set them manually.
However, I met some problems when I try to resolve this...

Comment (duckmoll) [reply NewDawn/NewDawn/MainPageViewController.swift]: This PR is getting me crazy.

Comment (tangziyi001) [reply NewDawn/NewDawn/MainPageViewController.swift]: What's the other problems?

Comment (duckmoll) [reply NewDawn/NewDawn/MainPageViewController.swift]: It is resolved.

**Comment by duckmoll** [comment]
Now this one is working.
